compiler/parser/RdrHsSyn.hs

```diff
@@ -2868,35 +2875,56 @@ mkChoiceInstanceDecl templateName CombinedChoiceData { ccdChoiceData = ChoiceDat
 
 mkInterfaceFixedChoiceInstanceDecl :: Located RdrName -> InterfaceChoiceSignature -> [LHsDecl GhcPs]
 mkInterfaceFixedChoiceInstanceDecl tycon InterfaceChoiceSignature {..} =
-  [ mkInstance "HasToAnyChoice" simpleContext (mkPrimMethod "_toAnyChoice" "EToAnyChoice")
-  , mkInstance "HasFromAnyChoice" simpleContext (mkPrimMethod "_fromAnyChoice" "EFromAnyChoice")
-  , mkInstance "HasExercise" exerciseContext (mkTemplateClassMethod "exercise" [cid, arg]
-      (mkPrimitive "primitive" "UExerciseInterface"
-        `mkApp` (mkParExpr $ mkQualVar (mkVarOcc "toInterfaceContractId")
-                  `mkAppType` rdrNameToType tycon
-                  `mkApp` mkUnqualVar (mkVarOcc "cid"))
-        `mkApp` (mkUnqualVar (mkVarOcc "arg"))
-        `mkApp` (mkParExpr $ mkQualVar (mkVarOcc "_typeRepForInterfaceExercise")
-                  `mkApp` mkUnqualVar (mkVarOcc "cid")))
-      Nothing)
+  [ mkInstance "HasToAnyChoice"
+      [paramType, choiceType, returnType]
+      simpleContext
+      (mkPrimMethod "_toAnyChoice" "EToAnyChoice")
+  , mkInstance "HasFromAnyChoice"
+      [paramType, choiceType, returnType]
+      simpleContext
+      (mkPrimMethod "_fromAnyChoice" "EFromAnyChoice")
+  , mkInstance "HasExerciseInterface"
+      [ifaceType, paramType, choiceType, returnType]
+      exerciseContext
+      (mkTemplateClassMethod "exerciseInterface" [pred, cid, arg]
+        (mkPrimitive "primitive" "UExerciseInterface"
+          `mkApp` (mkParExpr $ mkQualVar (mkVarOcc "toInterfaceContractId")
+                    `mkAppType` ifaceType
+                    `mkApp` mkUnqualVar (mkVarOcc "cid"))
+          `mkApp` (mkUnqualVar (mkVarOcc "arg"))
+          `mkApp` (mkParExpr $ mkQualVar (mkVarOcc "_typeRepForInterfaceExercise")
+                    `mkApp` mkUnqualVar (mkVarOcc "cid"))
+          `mkApp` (mkUnqualVar (mkVarOcc "pred")))
+        Nothing)
+  , mkInstance "HasExercise"
+      [paramType, choiceType, returnType]
+      exerciseContext
```

Comment:
I think I'd prefer to have `HasExerciseInterface Interface t Choice Result` as the instance context, but that results in an undecidable instance so I simply use the same context for both `Has{Exercise,}Interface` instances